atomic_reactor/plugins/pre_fetch_maven_artifacts.py

```diff
@@ -0,0 +1,216 @@
+    is_allowed_to_fail = False
+
+    NVR_REQUESTS_FILENAME = 'fetch-artifacts-koji.yaml'
+    URL_REQUESTS_FILENAME = 'fetch-artifacts-url.yaml'
+
+    DOWNLOAD_DIR = 'artifacts'
+
+    def __init__(self, tasker, workflow, koji_hub, koji_root,
+                 koji_proxyuser=None, koji_ssl_certs_dir=None,
+                 koji_krb_principal=None, koji_krb_keytab=None,
+                 allowed_domains=None):
+        """
+        :param tasker: DockerTasker instance
+        :param workflow: DockerBuildWorkflow instance
+        :param koji_hub: str, koji hub (xmlrpc)
+        :param koji_root: str, koji root (storage)
+        :param koji_proxyuser: str, proxy user
+        :param koji_ssl_certs_dir: str, path to "cert", "ca", and "serverca"
+        :param koji_krb_principal: str, name of Kerberos principal
+        :param koji_krb_keytab: str, Kerberos keytab
+        :param allowed_domains: list<str>: list of domains that are
+               allowed to be used when fetching artifacts by URL (case insensitive)
+        """
+        super(FetchMavenArtifactsPlugin, self).__init__(tasker, workflow)
+        koji_auth = {
+            'proxyuser': koji_proxyuser,
+            'ssl_certs_dir': koji_ssl_certs_dir,
+            'krb_principal': koji_krb_principal,
+            'krb_keytab': koji_krb_keytab,
+        }
+        # Remove empty values from auth dict to avoid login when not needed
+        koji_auth = {k: v for k, v in koji_auth.items() if v}
+        self.koji_info = {
+            'hub': koji_hub,
+            'root': koji_root,
+            'auth': koji_auth or None
+        }
+        self.path_info = PathInfo(topdir=self.koji_info['root'])
+        self.allowed_domains = set(domain.lower() for domain in allowed_domains or [])
+        self.workdir = self.workflow.source.get_dockerfile_path()[1]
+
+    def read_nvr_requests(self):
+        file_path = os.path.join(self.workdir, self.NVR_REQUESTS_FILENAME)
+        if not os.path.exists(file_path):
+            self.log.debug('%s not found', self.NVR_REQUESTS_FILENAME)
+            return []
+
+        nvr_requests = util.read_yaml(file_path, default=[],
+                                      schema='schemas/fetch-artifacts-nvr.json')
+        return [NvrRequest(**nvr_request) for nvr_request in nvr_requests]
+
+    def read_url_requests(self):
+        file_path = os.path.join(self.workdir, self.URL_REQUESTS_FILENAME)
+        if not os.path.exists(file_path):
+            self.log.debug('%s not found', self.URL_REQUESTS_FILENAME)
+            return []
+
+        return util.read_yaml(file_path, default=[],
+                              schema='schemas/fetch-artifacts-url.json')
+
+    def process_by_nvr(self, nvr_requests):
+        download_queue = []
+
+        for nvr_request in nvr_requests:
+            build_info = self.session.getBuild(nvr_request.nvr)
+            if not build_info:
+                raise ValueError('Build {} not found.'.format(nvr_request.nvr))
```

Comment:
We definitely should fail on error as currently. Just might be nice to report all the errors, then fail on one of them (the first?).